server/ocsp.go

```diff
@@ -400,11 +417,52 @@ func (srv *Server) NewOCSPMonitor(tc *tls.Config) (*tls.Config, *OCSPMonitor, er
 			}, nil
 		}
 
-		// GetClientCertificate returns a certificate that's presented to a
-		// server.
-		tc.GetClientCertificate = func(info *tls.CertificateRequestInfo) (*tls.Certificate, error) {
-			return &cert, nil
+		// Check whether need to verify staples from a client connection depending on the type.
+		switch kind {
+		case typeStringMap[ROUTER], typeStringMap[GATEWAY], typeStringMap[LEAF]:
+			tc.VerifyConnection = func(s tls.ConnectionState) error {
+				oresp := s.OCSPResponse
+				if oresp == nil {
+					return fmt.Errorf("missing OCSP Staple")
+				}
+
+				// Client route connections will verify the response of
+				// the staple.
+				if len(s.VerifiedChains) == 0 {
+					return fmt.Errorf("missing TLS verified chains")
+				}
+				chain := s.VerifiedChains[0]
+				resp, err := ocsp.ParseResponseForCert(oresp, chain[0], issuer)
+				if err != nil {
+					return fmt.Errorf("failed to parse OCSP response: %w", err)
+				}
+				if err := resp.CheckSignatureFrom(issuer); err != nil {
+					return err
+				}
+				if resp.Status != ocsp.Good {
+					return fmt.Errorf("bad status for OCSP Staple")
+				}
+
+				return nil
+			}
+
+			// When server makes a client connection, need to also present an OCSP Staple.
+			tc.GetClientCertificate = func(info *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+				raw, _, err := mon.getStatus()
+				if err != nil {
+					return nil, err
+				}
+				cert.OCSPStaple = raw
```

Comment:
We have to do this so that when connecting to a server, a client sends its staple which will be verified by the server attempting the connection /cc @philpennock